spec:ABCI2.0 - Uncommenting content related to vote xtensions and finalize block (#421)
cometbft/cometbft: spec/abci/abci++_app_requirements.md

Process *p*'s prepared proposal can differ in two different rounds where *p* is the proposer.

* Requirement 1 [`PrepareProposal`, timeliness]: If the network is in a synchronous period while processes 
  *p* and *q* are in *r<sub>p</sub>*,
  then the value of *TimeoutPropose* at *q* must be such that *q*'s propose timer does not time out
  (which would result in *q* prevoting `nil` in *r<sub>p</sub>*).

`PrepareProposal` stands on CometBFT's critical path. Thus,
Requirement 1 ensures the Application or operator will set a value for `TimeoutPropose` such that the time it takes
to fully execute `PrepareProposal` does not interfere with CometBFT's propose timer.

> Note that violation of Requirement 1 may just lead to further rounds, but will not compromise correctness.

Reviewer: I'd suggest `safety` instead of correctness, then, as using the wrong timeout cannot cause processes to disagree or violate validity. `Liveness`, however, seems incorrect as timeouts can lead to infinitely many extra rounds in which nil is pre-voted.

> `PrepareProposal` stands on CometBFT's critical path. Thus,
> Requirement 1 ensures the Application or operator will set a value for `TimeoutPropose` such that the time it takes
> to fully execute `PrepareProposal` does not interfere with CometBFT's propose timer.
> Note that violation of Requirement 1 may just lead to further rounds, but will not compromise correctness.

Reviewer: Discussing with @sergio-mena , we reached the following alternative text for line 34 (GitHub wouldn't let me add it as a suggestions.

```suggestion
Note that the violation of Requirement 1 may lead to further rounds, but will not compromise liveness because even though `TimeoutPropose` is used as the initial value for proposal timeouts, CometBFT will be dynamically adjust these timeouts such that they will eventually be enough for completing `PrepareProposal`.
```

> Note that violation of Requirement 1 may just lead to further rounds, but will not compromise correctness.

Reviewer: Discussing with @sergio-mena , we reached the following alternative text for **line 34** (GitHub wouldn't let me add it as a suggestion).

```
Note that the violation of Requirement 1 may lead to further rounds, but will not 
compromise liveness because even though `TimeoutPropose` is used as the initial
value for proposal timeouts, CometBFT will be dynamically adjust these timeouts
such that they will eventually be enough for completing `PrepareProposal`.
```

Requirement 1 is particularly important if *p*'s Application fully executes prepared blocks in `PrepareProposal`

> as a form of optimistic execution.

Reviewer 2: ```suggestion
as a form of immediate execution.
```

Reviewer: I believe that optimistic is more accurate here, because it signals that the execution may be wasted, while immediate does not. It is also a term commonly used for similar execution of transactions in DBs. I understand that we have used immediate in some other places so far, though, so would need to review them.

Author: I tend to agree with Lasaro, immediate sounds like the execution is final and the state change is persisted.

a correct process can broadcast at most one precommit
message in round *r*, height *h*.

 
Since, as stated in the [Methods](./abci++_methods.md#extendvote) section, `ResponseExtendVote`
is only called when the consensus algorithm
is about to broadcast a non-`nil` precommit message, a correct process can only produce one vote extension
in round *r*, height *h*.
Let *e<sup>r</sup><sub>p</sub>* be the vote extension that the Application of a correct process *p* returns via
`ResponseExtendVote` in round *r*, height *h*.
Let *w<sup>r</sup><sub>p</sub>* be the proposed block that *p*'s CometBFT passes to the Application via `RequestExtendVote`

> inround *r*, height *h*. 

Reviewer: ```suggestion
in round *r*, height *h*.
```

still received in sequence.
 -->

#### FinalizeBlock

When the consensus algorithm decides on a block, CometBFT uses `FinalizeBlock` to send the

> decided block's data to the Application, which uses it to transition its state, but not persist it;
> persisting will be done during `Commit`.

Reviewer 2: ```suggestion
decided block's data to the Application, which uses it to transition its state, but MUST NOT persist it;
persisting MUST be done during `Commit`.
```
I'm suggesting this now because this helps in tackling #203

### Transaction Results

The Application is expected to return a list of
[`ExecTxResult`](./abci%2B%2B_methods.md#exectxresult) in
[`ResponseFinalizeBlock`](./abci%2B%2B_methods.md#finalizeblock). The list of transaction

> results must respect the same order as the list of transactions delivered via

Reviewer 2: ```suggestion
results MUST respect the same order as the list of transactions delivered via
```

used to explicitly prioritize transactions in the mempool for inclusion in a block
proposal.

#### Specifics of `ExecTxResult`

> `FinalizeBlock` is the workhorse of the blockchain.CometBFT delivers the decided block, 

Reviewer: ```suggestion
`FinalizeBlock` is the workhorse of the blockchain. CometBFT delivers the decided block,
```
